packages/headless/doc-parser/src/entity-sorter.ts

```diff
@@ -0,0 +1,57 @@
+import {AnyEntity, FuncEntity, isFunctionEntity} from './entity';
+
+export function sortEntities(entities: AnyEntity[]) {
+  const [methods, attributes] = mapIntoGroups<
+    AnyEntity,
+    FuncEntity,
+    Exclude<AnyEntity, FuncEntity>
+  >(entities, (entity, addMethod, addAttribute) =>
+    isFunctionEntity(entity) ? addMethod(entity) : addAttribute(entity)
+  );
+
+  const [optionalAttributes, mandatoryAttributes] = mapIntoGroups(
+    attributes,
+    (entity, addOptionalAttribute, addMandatoryAttribute) =>
+      entity.isOptional
+        ? addOptionalAttribute(entity)
+        : addMandatoryAttribute(entity)
+  );
+
+  const sortedMandatoryAttributes = alphabeticallySortEntities(
+    mandatoryAttributes
+  );
+  const sortedOptionalAttributes = alphabeticallySortEntities(
+    optionalAttributes
+  );
+  const sortedMethods = alphabeticallySortEntities(methods);
+
+  return [
+    ...sortedMandatoryAttributes,
+    ...sortedOptionalAttributes,
+    ...sortedMethods,
+  ];
+}
+
+function mapIntoGroups<T, A = T, B = T>(
```

Comment:
This function isn't particularly useful anymore. In fact, it's pretty much equivalent to a `reduce`. However, its purpose is clear and using it makes the code a lot more readable.